### ISSUE TITLE
Reconciliation is not re-queued based on the `credentialsRotationPolicy` config

### ISSUE THREAD

**Issue by pbusko** [SAP/sap-btp-service-operator#253]
Currently, the `credentialsRotationPolicy` accepts any valid `time.Duration`, but the reconciliation loop doesn't re-queue within the duration specified.

For example, if we create a binding with the following `credentialsRotationPolicy`:

```yaml
....
    credentialsRotationPolicy:
      enabled: true
      rotationFrequency: "1h"
      rotatedBindingTTL: "30m"
```

The credentials are not rotated for the next 10 hours (default full reconciliation cycle). On top of that, the backup binding won't be cleaned up after specified `30m` duration, but rather only during the following full reconciliation (~ 20 hours from the time of initial creation).

One potential solution, is to exit the `Reconcile` loop with the `RequeueAfter` option set, making the reconciliation loop to start at the configured interval. For example:

```golang
return ctrl.Result{Requeue: true, RequeueAfter: getNextRotationTime(binding)}, nil
```

**Comment by c0d1ngm0nk3y** [comment]
And please also rethink/adapt the [documentation](https://github.com/SAP/sap-btp-service-operator#credentials-rotation) on this. Right now, it is simply wrong, but even when this gets fixed, having times in `ms` or even `ns` seems wrong.

**Comment by c0d1ngm0nk3y** [comment]
@evyaffe or @kerenlahav Any chance to get some comment?

**Comment by kerenlahav** [comment]
As mentioned in the documentation, the rotation will occur during a full reconciliation loop, this means the minimum time for rotation is 10 hours, which is not a long time in productive flows. For  testing purposes it is possible to trigger the rotation manually by adding an annotation **services.cloud.sap.com/forceRotate**

**Comment by kerenlahav** [comment]
> And please also rethink/adapt the [documentation](https://github.com/SAP/sap-btp-service-operator#credentials-rotation) on this. Right now, it is simply wrong, but even when this gets fixed, having times in `ms` or even `ns` seems wrong.

Hi, can you please elaborate on what is wrong?
regarding the "ns ms", the minimum time is 10 hours so it doesnt matter if you set the value in ns, I will add it to the documentation

**Comment by c0d1ngm0nk3y** [comment]
> Hi, can you please elaborate on what is wrong?

Maybe it is not "wrong" per se - I am no lawyer - but it is extremely misleading to say the least :)

> regarding the "ns ms", the minimum time is 10 hours so it doesnt matter if you set the value in ns, I will add it to the documentation

Exactly. The documentation mentions "ns" and "ms" which are pointless and doesn't even mention that they are pointless. You could argue for "hours", but anything below that is not even worth documenting because it is a "lie" anyway. Wouldn't you agree?

**Comment by c0d1ngm0nk3y** [comment]
> For testing purposes it is possible to trigger the rotation manually by adding an annotation **services.cloud.sap.com/forceRotate**

Thanks for pointing to this. Just to clarify since the documentation is a bit vague on this. This will invalidate the existing credentials immediately, right? The name implies it for sure, but I wanted to recheck.

**Comment by kerenlahav** [comment]
I agree the documentation is not that clear on this topic, i'll take it with the TW

_"Thanks for pointing to this. Just to clarify since the documentation is a bit vague on this. This will invalidate the existing credentials immediately, right? The name implies it for sure, but I wanted to recheck."_
It will rotate and invalidate according to the configuration, the annotation is just to trigger an event on the resource.

**Comment by loewenstein** [comment]
> It will rotate and invalidate according to the configuration, the annotation is just to trigger an event on the resource.

But that's a problem, isn't it? The most likely use case I can imagine for a **force** rotation is a credential leak. That's where you wouldn't want to wait for up to 10 hours until the next full reconciliation happens.

Besides, an implementation for timely rotation, i.e. as configured, shouldn't be very difficult or complex with the requeue capabilities of the controller runtime.

**Comment by c0d1ngm0nk3y** [comment]
> The most likely use case I can imagine for a **force** rotation is a credential leak.

What should be done in this case? I also thought `forceRotate` is exactly for that case. Again, the documentation did not say anything on that important topic, or at least I haven't found it.

@kerenlahav Since this is a different issue to this one (but related), should I open a separate issue?

**Comment by kerenlahav** [comment]
I checked the code and what I wrote is incorrect, using the forceRotate will create a new binding and will store the old binding (according to the TTL) regardless of the rotation configuration.

Regarding the credential leak, you should have a maximum of 2 bindings (current and old with TTL), why do you need an annotation to delete them? instead of adding an annotation you can delete them directly, no?

I agree that the credentials rotation process can be improved, we'll create a BLI and consider the retryAfter mechanism.
Just for us to better understand the use case, can you elaborate on why you need the retrtyAfter? is it for testing purposes? is the invalidate/rotate time must be accurate?

**Comment by loewenstein** [comment]
> why do you need an annotation to delete them? instead of adding an annotation you can delete them directly, no?

It's the service credentials that need to be invalidated on the service side, so the communication with the service broker via the service manager is the important bit for emergency rotation because of a credential leak I would say.

> I agree that the credentials rotation process can be improved, we'll create a BLI and consider the retryAfter mechanism.
> Just for us to better understand the use case, can you elaborate on why you need the retryAfter? is it for testing purposes? is the invalidate/rotate time must be accurate?

Testing was merely how we discovered the discrepancy between documentation/configuration and the implementation.
Rotation frequencies below the 10 hours might be unnecessary for productive cases (not sure but might), so I understand the current choice of implementation. However, allowing configuration that is known to deliver surprising results is just generally something that's worth avoiding.

The two main realistic use cases I can see are testing and trying out. The latter might turn out to be the most troublesome, because if I configure a one hour rotation and don't see it happen, I am less likely to trust the solution in the end.

Also, I think the `forceRotation` use case will anyway require the re-queue behaviour, so a consistent implementation of the regular rotation doesn't seem to add much. Maybe checking the configuration and disallow anything below ~5 minutes...

**Comment by kerenlahav** [comment]
> It's the service credentials that need to be invalidated on the service side, so the communication with the service broker via the service manager is the important bit for emergency rotation because of a credential leak I would say.

When you delete a service binding (rotated or not) the operator sends a delete request to the relevant broker through Service Manager, it will be deleted from the cluster only after a SUCCESS response was returned from the broker. (you should delete the binding, not the secret)



> Also, I think the `forceRotation` use case will anyway require the re-queue behaviour, so a consistent implementation of the regular rotation doesn't seem to add much. Maybe checking the configuration and disallow anything below ~5 minutes...

why will it require a re-queue? forceRotation will trigger the rotation immediately

> Testing was merely how we discovered the discrepancy between documentation/configuration and the implementation.
> Rotation frequencies below the 10 hours might be unnecessary for productive cases (not sure but might), so I understand the current choice of implementation. However, allowing configuration that is known to deliver surprising results is just generally something that's worth avoiding.
I don't think testing should affect productive code, there are other solutions to simulate the rotation but I will take it with the PO since it is a product decision
